File: EssentialAchievements/src/main/java/essential/achievements/Achievement.java
package essential.achievements;

import arc.Events;
import essential.database.data.PlayerData;
import mindustry.Vars;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.file.Files;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

import static essential.database.data.AchievementDataKt.hasAchievement;
import static essential.database.data.AchievementDataKt.setAchievement;

public enum Achievement {
    // int array values are current value and target value
    CrawlerBlockDestroyer {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.crawler.block.destroy", "0"));
        }
    },

    SoloMapClear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.solo", "0"));
        }
    },

    LongPlayNoAfk {
        @Override
        public int value() {
            return 18000; // 5 hours in seconds
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.time.noafk", "0"));
        }
    },

    NoMiningClear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.nomining", "0"));
        }
    },

    SerpuloPvPWin {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.win.serpulo", "0"));
        }
    },

    ErekirPvPWin {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.win.erekir", "0"));
        }
    },

    BothPlanetsPvPWin {
        @Override
        public int value() {
            return 10;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.win.both", "0"));
        }
    },

    LeaveAndLosePvP {
        @Override
        public int value() {
            return 10;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.leave.lose", "0"));
        }
    },

    PvPWinMaster {
        @Override
        public int value() {
            return 300;
        }

        @Override
        public int current(PlayerData data) {
            return data.getPvpWinCount();
        }
    },

    MapClearMaster {
        @Override
        public int value() {
            return 10;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.count", "0"));
        }
    },

    TurretMultiKill {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.turret.multikill", "0"));
        }
    },

    QuillKiller {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.turret.quill.kill", "0"));
        }
    },

    ZenithKiller {
        @Override
        public int value() {
            return 30;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.turret.zenith.kill", "0"));
        }
    },

    OmuraHorizonKiller {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.omura.horizon.kill", "0"));
        }
    },

    PvPWinStreak {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.win.streak", "0"));
        }
    },

    PvPWinRate {
        @Override
        public int value() {
            return 20;
        }

        @Override
        public int current(PlayerData data) {
            int result;
            try {
                int total = data.getPvpWinCount() + data.getPvpLoseCount();
                if (total < 10) { // Require at least 10 games for win rate calculation
                    result = 0;
                } else {
                    result = data.getPvpWinCount() * 100 / total;
                }
            } catch (ArithmeticException e) {
                result = 0;
            }
            return result;
        }
    },

    PvPUnderdog {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.underdog", "0"));
        }
    },

    WarpServerDisconnect {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.warp.disconnect", "0"));
        }
    },

    ExplosionKiller {
        @Override
        public int value() {
            return 10;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.explosion.kill", "0"));
        }
    },

    LowPowerClear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.lowpower", "0"));
        }
    },

    PvPDefeatStreak {
        @Override
        public int value() {
            return 5;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.defeat.streak", "0"));
        }
    },

    NoPowerClear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.nopower", "0"));
        }
    },

    NoTurretsClear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.noturrets", "0"));
        }
    },

    DuoTurretSurvival {
        @Override
        public int value() {
            return 100;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.wave.duo", "0"));
        }
    },

    FlareOnlyClear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.flareonly", "0"));
        }
    },

    VotingBan {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.voting.ban", "0"));
        }
    },

    PvPContribution {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.pvp.contribution", "0"));
        }
    },

    APM50 {
        @Override
        public int value() {
            return 50;
        }

        @Override
        public int current(PlayerData data) {
            return data.getApm();
        }
    },

    APM100 {
        @Override
        public int value() {
            return 100;
        }

        @Override
        public int current(PlayerData data) {
            return data.getApm();
        }
    },

    APM200 {
        @Override
        public int value() {
            return 200;
        }

        @Override
        public int current(PlayerData data) {
            return data.getApm();
        }
    },

    MapProvider {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.provider", "0"));
        }
    },

    FeedbackProvider {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.feedback.provider", "0"));
        }
    },

    Builder {
        @Override
        public int value() {
            return 100000;
        }

        @Override
        public int current(PlayerData data) {
            return data.getBlockPlaceCount();
        }
    },

    Deconstructor {
        @Override
        public int value() {
            return 100000;
        }

        @Override
        public int current(PlayerData data) {
            return data.getBlockBreakCount();
        }
    },

    Creator {
        @Override
        public int value() {
            return 360000;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.time.sandbox", "0"));
        }
    },

    Eliminator {
        @Override
        public int value() {
            return 100;
        }

        @Override
        public int current(PlayerData data) {
            return data.getPvpWinCount();
        }
    },

    Defender {
        @Override
        public int value() {
            return 10000;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.wave", "0"));
        }
    },

    Aggressor {
        @Override
        public int value() {
            return 50;
        }

        @Override
        public int current(PlayerData data) {
            return data.getAttackClear();
        }
    },

    Serpulo {
        @Override
        public int value() {
            return 360000;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.time.serpulo", "0"));
        }
    },

    Erekir {
        @Override
        public int value() {
            return 360000;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.time.erekir", "0"));
        }
    },

    TurbidWater {
        @Override
        public int value() {
            return 360000;
        }

        @Override
        public int current(PlayerData data) {
            return data.getTotalPlayed();
        }
    },

    BlackWater {
        @Override
        public int value() {
            return 720000;
        }

        @Override
        public int current(PlayerData data) {
            return data.getTotalPlayed();
        }
    },

    Oil {
        @Override
        public int value() {
            return 1080000;
        }

        @Override
        public int current(PlayerData data) {
            return data.getTotalPlayed();
        }
    },

    SerpuloQuad {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.unit.serpulo.quad", "0"));
        }
    },

    Lord {
        @Override
        public int value() {
            return 70;
        }

        @Override
        public int current(PlayerData data) {
            int result;
            try {
                int total = data.getPvpWinCount() + data.getPvpLoseCount();
                if (total < 50) {
                    result = 0;
                } else {
                    result = data.getPvpWinCount() * 100 / total;
                }
            } catch (ArithmeticException e) {
                result = 0;
            }
            return result;
        }
    },

    Chatter {
        @Override
        public int value() {
            return 10000;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.time.chat", "0"));
        }
    },

    // ??
    MeetOwner {
        @Override
        public int value() {
            return 1;
        }

        private final boolean isHidden = true;

        @Override
        public boolean isHidden() {
            return isHidden;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.time.meetowner", "0"));
        }
    },

    // Specific map clear achievements
    Asteroids {
        @Override
        public int value() {
            return 1;
        }

        private final boolean isHidden = true;

        @Override
        public boolean isHidden() {
            return isHidden;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.asteroids", "0"));
        }

        @Override
        public boolean success(PlayerData data) {
            String mapHash = "7b032cc7815022be644d00a877ae0388";
            if (Achievement.getMapHash().equals(mapHash)) {
                data.getStatus().put("record.map.clear.asteroids", "1");
                return true;
            } else {
                return false;
            }
        }
    },

    Transcendence {
        @Override
        public int value() {
            return 1;
        }

        private final boolean isHidden = true;

        @Override
        public boolean isHidden() {
            return isHidden;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.transcendence", "0"));
        }

        @Override
        public boolean success(PlayerData data) {
            String mapHash = "f355b3d91d5d8215e557ff045b3864ef";
            if (Achievement.getMapHash().equals(mapHash)) {
                data.getStatus().put("record.map.clear.transcendence", "1");
                return true;
            } else {
                return false;
            }
        }
    },

    NewYear {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.chat.newyear", "0"));
        }
    },

    Loyal {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.login.loyal", "0"));
        }
    },

    WaterExtractor {
        @Override
        public int value() {
            return 10;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.build.waterextractor", "0"));
        }
    },

    Attendance {
        @Override
        public int value() {
            return 100;
        }

        @Override
        public int current(PlayerData data) {
            return data.getAttendanceDays();
        }
    },

    LoyalSixMonths {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.login.loyal.sixmonths", "0"));
        }
    },

    LoyalOneYearSixMonths {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.login.loyal.oneyearsixmonths", "0"));
        }
    },

    AllMaps {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return Integer.parseInt(data.getStatus().getOrDefault("record.map.clear.all", "0"));
        }
    },

    DiscordAuth {
        @Override
        public int value() {
            return 1;
        }

        @Override
        public int current(PlayerData data) {
            return data.getDiscordID() != null && !data.getDiscordID().isEmpty() ? 1 : 0;
        }
    };

    public abstract int value();

    public boolean isHidden() {
        return false;
    }

    public abstract int current(PlayerData data);

    public boolean success(PlayerData data) {
        // Prevent achievements from being cleared in sandbox mode
        if (Vars.state.rules.infiniteResources) {
            return false;
        }

        String achievementName = this.toString().toLowerCase(Locale.getDefault());

        if (hasAchievement(data, achievementName)) {
            return false;
        } else {
            return current(data) >= value();
        }
    }

    public void set(PlayerData data) {
        String achievementName = this.toString().toLowerCase(Locale.getDefault());

        // Store in status for temporary use during this session
        if (!data.getStatus().containsKey("achievement." + achievementName)) {
            data.getStatus().put(
                "achievement." + achievementName,
                LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            );

            setAchievement(data, achievementName);

            Events.fire(new CustomEvents.AchievementClear(this, data));
        }
    }

    private static String getMapHash() {
        try {
            byte[] data = Files.readAllBytes(Vars.state.map.file.file().toPath());
            byte[] hash = MessageDigest.getInstance("MD5").digest(data);
            return new BigInteger(1, hash).toString(16);
        } catch (NoSuchAlgorithmException | IOException e) {
            return "";
        }
    }
}

File: EssentialAchievements/src/main/java/essential/achievements/Achievement.kt
package essential.achievements

/*
public enum Achievement {
   Builder, Deconstructor,
   // Specific
   Creator, Eliminator, Defender, Aggressor, Serpulo, Erekir,
   // Time
   TurbidWater, BlackWater, Oil,
   // PvP
   Lord,
   */
/*Iron, Bronze, Silver, Gold, Platinum, Master, GrandMaster*/ /*

    // Wave
    // Attack
}*/

import arc.Events
import essential.core.Main.Companion.scope
import essential.database.data.PlayerDataEntity
import essential.database.data.hasAchievement
import essential.database.data.setAchievement
import kotlinx.coroutines.launch
import mindustry.Vars
import java.io.IOException
import java.math.BigInteger
import java.nio.file.Files
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*

enum class Achievement {
    // int 배열값은 현재 값과 목표 값
    CrawlerBlockDestroyer {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.crawler.block.destroy", "0").toInt()
        }
    },

    SoloMapClear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.solo", "0").toInt()
        }
    },

    LongPlayNoAfk {
        override fun value(): Int{
            return 18000 // 5 hours in seconds
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.time.noafk", "0").toInt()
        }
    },

    NoMiningClear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.nomining", "0").toInt()
        }
    },

    SerpuloPvPWin {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.win.serpulo", "0").toInt()
        }
    },

    ErekirPvPWin {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.win.erekir", "0").toInt()
        }
    },

    BothPlanetsPvPWin {
        override fun value(): Int{
            return 10
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.win.both", "0").toInt()
        }
    },

    LeaveAndLosePvP {
        override fun value(): Int{
            return 10
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.leave.lose", "0").toInt()
        }
    },

    PvPWinMaster {
        override fun value(): Int{
            return 300
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.pvpWinCount.toInt()
        }
    },

    MapClearMaster {
        override fun value(): Int{
            return 10
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.count", "0").toInt()
        }
    },

    TurretMultiKill {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.turret.multikill", "0").toInt()
        }
    },

    QuillKiller {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.turret.quill.kill", "0").toInt()
        }
    },

    ZenithKiller {
        override fun value(): Int{
            return 30
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.turret.zenith.kill", "0").toInt()
        }
    },

    OmuraHorizonKiller {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.omura.horizon.kill", "0").toInt()
        }
    },

    PvPWinStreak {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.win.streak", "0").toInt()
        }
    },

    PvPWinRate {
        override fun value(): Int{
            return 20
        }

        override fun current(data: PlayerDataEntity): Int{
            var result: Int
            try {
                val total: Int = data.pvpWinCount + data.pvpLoseCount
                if (total < 10) { // Require at least 10 games for win rate calculation
                    result = 0
                } else {
                    result = data.pvpWinCount * 100 / total
                }
            } catch (e: ArithmeticException) {
                result = 0
            }
            return result
        }
    },

    PvPUnderdog {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.underdog", "0").toInt()
        }
    },

    WarpServerDisconnect {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.warp.disconnect", "0").toInt()
        }
    },

    ExplosionKiller {
        override fun value(): Int{
            return 10
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.explosion.kill", "0").toInt()
        }
    },

    LowPowerClear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.lowpower", "0").toInt()
        }
    },

    PvPDefeatStreak {
        override fun value(): Int{
            return 5
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.defeat.streak", "0").toInt()
        }
    },

    NoPowerClear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.nopower", "0").toInt()
        }
    },

    NoTurretsClear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.noturrets", "0").toInt()
        }
    },

    DuoTurretSurvival {
        override fun value(): Int{
            return 100
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.wave.duo", "0").toInt()
        }
    },

    FlareOnlyClear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.flareonly", "0").toInt()
        }
    },

    VotingBan {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.voting.ban", "0").toInt()
        }
    },

    PvPContribution {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.pvp.contribution", "0").toInt()
        }
    },

    APM50 {
        override fun value(): Int{
            return 50
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.apm
        }
    },

    APM100 {
        override fun value(): Int{
            return 100
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.apm
        }
    },

    APM200 {
        override fun value(): Int{
            return 200
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.apm
        }
    },

    MapProvider {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.provider", "0").toInt()
        }
    },

    FeedbackProvider {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.feedback.provider", "0").toInt()
        }
    },

    Builder {
        override fun value(): Int{
            return 100000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.blockPlaceCount
        }
    },
    Deconstructor {
        override fun value(): Int{
            return 100000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.blockBreakCount
        }
    },

    Creator {
        override fun value(): Int{
            return 360000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.time.sandbox", "0").toInt()
        }
    },
    Eliminator {
        override fun value(): Int{
            return 100
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.pvpWinCount.toInt()
        }
    },
    Defender {
        override fun value(): Int{
            return 10000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.wave", "0").toInt()
        }
    },
    Aggressor {
        override fun value(): Int{
            return 50
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.attackClear
        }
    },
    Serpulo {
        override fun value(): Int{
            return 360000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.time.serpulo", "0").toInt()
        }
    },
    Erekir {
        override fun value(): Int{
            return 360000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.time.erekir", "0").toInt()
        }
    },

    TurbidWater {
        override fun value(): Int{
            return 360000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.totalPlayed
        }
    },
    BlackWater {
        override fun value(): Int{
            return 720000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.totalPlayed
        }
    },
    Oil {
        override fun value(): Int{
            return 1080000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.totalPlayed
        }
    },

    SerpuloQuad {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.unit.serpulo.quad", "0").toInt()
        }
    },

    Lord {
        override fun value(): Int{
            return 70
        }

        override fun current(data: PlayerDataEntity): Int{
            var result: Int
            try {
                val total: Int = data.pvpWinCount + data.pvpLoseCount
                if (total < 50) {
                    result = 0
                } else {
                    result = data.pvpWinCount * 100 / total
                }
            } catch (e: ArithmeticException) {
                result = 0
            }
            return result
        }
    },

    Chatter {
        override fun value(): Int{
            return 10000
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.time.chat", "0").toInt()
        }
    },

    // ??
    MeetOwner {
        override fun value(): Int{
            return 1
        }

        override val isHidden = true

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.time.meetowner", "0").toInt()
        }
    },

    // Specific map clear achievements
    Asteroids {
        override fun value(): Int{
            return 1
        }

        override val isHidden = true

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.asteroids", "0").toInt()
        }

        override fun success(data: PlayerDataEntity): Boolean {
            val mapHash = "7b032cc7815022be644d00a877ae0388"
            if (Achievement.Companion.mapHash == mapHash) {
                data.status.put("record.map.clear.asteroids", "1")
                return true
            } else {
                return false
            }
        }
    },

    Transcendence {
        override fun value(): Int{
            return 1
        }

        override val isHidden = true

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.transcendence", "0").toInt()
        }

        override fun success(data: PlayerDataEntity): Boolean {
            val mapHash = "f355b3d91d5d8215e557ff045b3864ef"
            if (Achievement.Companion.mapHash == mapHash) {
                data.status.put("record.map.clear.transcendence", "1")
                return true
            } else {
                return false
            }
        }
    },

    NewYear {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.chat.newyear", "0").toInt()
        }
    },

    Loyal {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.login.loyal", "0").toInt()
        }
    },

    WaterExtractor {
        override fun value(): Int{
            return 10
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.build.waterextractor", "0").toInt()
        }
    },

    Attendance {
        override fun value(): Int{
            return 100
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.attendanceDays
        }
    },

    LoyalSixMonths {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.login.loyal.sixmonths", "0").toInt()
        }
    },

    LoyalOneYearSixMonths {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.login.loyal.oneyearsixmonths", "0").toInt()
        }
    },

    AllMaps {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return data.status.getOrDefault("record.map.clear.all", "0").toInt()
        }
    },

    DiscordAuth {
        override fun value(): Int{
            return 1
        }

        override fun current(data: PlayerDataEntity): Int{
            return if (data.discordID?.isNotEmpty() == true) 1 else 0
        }
    };

    abstract fun value(): Int
    open val isHidden: Boolean = false

    abstract fun current(data: PlayerDataEntity): Int
    open fun success(data: PlayerDataEntity): Boolean {
        // Prevent achievements from being cleared in sandbox mode
        if (mindustry.Vars.state.rules.infiniteResources) {
            return false
        }

        // First check if the achievement is already completed in status
        val achievementName = this.toString().lowercase(Locale.getDefault())
        val isCompletedInStatus = data.status.containsKey("achievement.$achievementName")

        // If it's already marked as completed in status, return true
        if (isCompletedInStatus) {
            return true
        }

        // Otherwise, check if the current value meets the target
        return current(data) >= value()
    }

    /**
     * Check if the achievement is completed in the database
     * This is a suspending function and should be called from a coroutine
     */
    suspend fun isCompletedInDatabase(data: PlayerDataEntity): Boolean {
        val achievementName = this.toString().lowercase(Locale.getDefault())
        return hasAchievement(data, achievementName)
    }

    fun set(data: PlayerDataEntity) {
        val achievementName = this.toString().lowercase(Locale.getDefault())

        // Store in status for temporary use during this session
        if (!data.status.containsKey("achievement.$achievementName")) {
            data.status.put(
                "achievement.$achievementName",
                LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            )

            // Also store in database for permanent storage
            scope.launch {
                setAchievement(data, achievementName)
            }

            Events.fire(CustomEvents.AchievementClear(this, data))
        }
    }

    companion object {
        private val mapHash: String?
            get() {
                try {
                    val data = Files.readAllBytes(Vars.state.map.file.file().toPath())
                    val hash = MessageDigest.getInstance("MD5").digest(data)
                    return BigInteger(1, hash).toString(16)
                } catch (e: NoSuchAlgorithmException) {
                    return ""
                } catch (e: IOException) {
                    return ""
                }
            }
    }
}
